shaders/pbr-frag.glsl

```diff
@@ -46,180 +59,161 @@ varying vec3 v_Position;
-vec3 lambertianDiffuse(PBRInfo pbrInputs)
-{
-  return pbrInputs.baseColor / M_PI;
-}
-
-// The following equations model the Fresnel reflectance term of the spec equation (aka F())
-// implementation of fresnel from “An Inexpensive BRDF Model for Physically based Rendering” by Christophe Schlick
-vec3 fresnelSchlick2(PBRInfo pbrInputs)
+// Helper function to get the tangent space matrix under several possible model configurations
+mat3 getTangentSpaceMatrix()
 {
-  return pbrInputs.reflectance0 + (pbrInputs.reflectance90 - pbrInputs.reflectance0) * pow(clamp(1.0 - pbrInputs.VdotH, 0.0, 1.0), 5.0);
-}
-
-// Simplified implementation of fresnel from “An Inexpensive BRDF Model for Physically based Rendering” by Christophe Schlick
-vec3 fresnelSchlick(PBRInfo pbrInputs)
-{
-  return pbrInputs.metalness + (vec3(1.0) - pbrInputs.metalness) * pow(1.0 - pbrInputs.VdotH, 5.0);
-}
+  #ifndef HAS_TANGENTS
+    vec3 pos_dx = dFdx(v_Position);
+    vec3 pos_dy = dFdy(v_Position);
+    vec3 tex_dx = dFdx(vec3(v_UV, 0.0));
+    vec3 tex_dy = dFdy(vec3(v_UV, 0.0));
+    vec3 t = (tex_dy.t * pos_dx - tex_dx.t * pos_dy) / (tex_dx.s * tex_dy.t - tex_dy.s * tex_dx.t);
+
+    #ifdef HAS_NORMALS
+      vec3 ng = normalize(v_Normal);
+    #else
+      vec3 ng = cross(pos_dx, pos_dy);
+    #endif
+
+    t = normalize(t - ng * dot(ng, t));
+    vec3 b = normalize(cross(ng, t));
+    mat3 tbn = mat3(t, b, ng);
+  #else // HAS_TANGENTS && HAS_NORMALS
+    mat3 tbn = v_TBN;
+  #endif
 
-// The following equations model the geometric occlusion term of the spec equation  (aka G())
-// Implementation from “A Reflectance Model for Computer Graphics” by Robert Cook and Kenneth Torrance,
-float geometricOcclusionCookTorrance(PBRInfo pbrInputs)
-{
-  return min(min(2.0 * pbrInputs.NdotV * pbrInputs.NdotH / pbrInputs.VdotH, 2.0 * pbrInputs.NdotL * pbrInputs.NdotH / pbrInputs.VdotH), 1.0);
+  return tbn;
 }
 
-// implementation of microfacet occlusion from “An Inexpensive BRDF Model for Physically based Rendering” by Christophe Schlick
-float geometricOcclusionSchlick(PBRInfo pbrInputs)
+// Calculation of the lighting contribution from an optional Image Based Light source.
+// Precomputed Environment Maps are required uniform inputs and are computed as outlined in [1].
+// See our README.md on Environment Maps [3] for additional discussion.
+vec3 getIBLContribution(PBRInfo pbrInputs, vec3 n, vec3 reflection)
 {
-  float k = pbrInputs.perceptualRoughness * 0.79788; // 0.79788 = sqrt(2.0/3.1415); perceptualRoughness = sqrt(alphaRoughness);
-  // alternately, k can be defined with
-  // float k = (pbrInputs.perceptualRoughness + 1) * (pbrInputs.perceptualRoughness + 1) / 8;
-
-  float l = pbrInputs.LdotH / (pbrInputs.LdotH * (1.0 - k) + k);
-  float n = pbrInputs.NdotH / (pbrInputs.NdotH * (1.0 - k) + k);
-  return l * n;
+    float mipCount = 9.0; // resolution of 512x512
```

Comment:
There is a mix of 4 spaces and 2 spaces for tabs right now. The original file used 2 spaces, but it might be better to change all tabs to 4 spaces instead for better readability? Not sure how easy it is to do that with an editor.